### PR TITLE
[Misc] LoRA + Chunked Prefill

### PR DESCRIPTION
**Pull request by aurickq**
Make LoRA work with chunked prefill, taking the same approach as https://github.com/vllm-project/vllm/pull/4994 but updated to latest code.

Also needed to modify the request sorting by the scheduler so prefill sequences always precede decoding sequences.

FIX #xxxx (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by NickLucche** [comment]
I believe putting *all* prefill requests first will break the Prefill/Decode ratio assumption as introduced here https://arxiv.org/abs/2308.16369, favoring memory-bound operations ends up in decreased throughput.

I am afraid I can't gather a lot of context from this PR, why is it needed to put prefills first here (converging to the default non-chunked scheduler policy)?

**Comment by aurickq** [comment]
> I believe putting _all_ prefill requests first will break the Prefill/Decode ratio assumption as introduced here https://arxiv.org/abs/2308.16369, favoring memory-bound operations ends up in decreased throughput.
> 
> I am afraid I can't gather a lot of context from this PR, why is it needed to put prefills first here (converging to the default non-chunked scheduler policy)?

My top-level comment was a bit ambiguous. The sorting is applied to the sequences within each batch, not in the scheduling queue, so it should not affect the ability to batch together prefill and decode sequences.

File: tests/lora/test_chatglm3.py [hunk 1]
```diff
@@ -43,7 +43,8 @@ def test_chatglm3_lora(chatglm3_lora_files):
                    enable_lora=True,
                    max_loras=4,
                    max_lora_rank=64,
-                   trust_remote_code=True)
+                   trust_remote_code=True,
+                   enable_chunked_prefill=True)
```
Comment (rkooo567) [inline tests/lora/test_chatglm3.py:47]: can you create at least 1 test to test it without chunked prefill?

File: vllm/core/scheduler.py [hunk 2]
```diff
@@ -152,7 +152,8 @@ def is_empty(self) -> bool:
     def _sort_by_lora_ids(self):
         self.scheduled_seq_groups = sorted(
             self.scheduled_seq_groups,
-            key=lambda g: (g.seq_group.lora_int_id, g.seq_group.request_id))
+            key=lambda g: (not g.seq_group.is_prefill(), g.seq_group.
```
Comment (rkooo567) [inline vllm/core/scheduler.py:155]: can you tell me the motivation of this change?

**Review by rkooo567 — APPROVED** [review]
nit comments!

File: vllm/core/scheduler.py (hunk 2 above)
Comment (aurickq) [reply vllm/core/scheduler.py:155]: when chunked prefill is disabled, i guess `g.seq_group.is_prefill()` will all be the same value, so it wouldn't change the ordering?

Comment (rkooo567) [reply vllm/core/scheduler.py:155]: ah I see. yeah I think that's right

**Comment by mces89** [comment]
@aurickq Thanks for this pr, i'm trying to test it on my own data. for my model, i set a large max_model_len and small max-num-batched-tokens. then i get error:
 "ValueError: The model's max seq len (100000) is larger than the maximum number of tokens that can be stored in KV cache (92240). Try increasing `gpu_memory_utilization` or decreasing `max_model_len` when initializing the engine."
 
 I'm using both the --enable-lora and --enable-chunked-prefill. My understanding is that as i set a small max-num-batched-tokens, for chunked-prefill, we just need to check max-num-batched-tokens'scomparison with the kv cache instead of max_model_len, why we are still using max_model_len to check?

**Comment by mces89** [comment]
Hi, Can I ask when this pr can be merged so we can use this lora+chunked_prefill as soon as possible.

**Comment by kouroshHakha** [comment]
Hi @aurickq , awesome that you are working on this. Do you have an estimate on when this will be pushed over to finish line? Is this waiting on review or is there more blockers?

**Comment by aurickq** [comment]
@mces89 @kouroshHakha I rebased and all the tests are passing, just waiting for someone to merge it now. Perhaps @rkooo567 @simon-mo ?

**Comment by jeejeelee** [comment]
@aurickq  Thanks for your contribution and patience

**Pull request merged**